pkg/controllers/allocation/batch.go

```diff
@@ -0,0 +1,178 @@
+func (b *Batcher) Wait(obj metav1.Object) {
+	waitBatchOp := &batchOp{kind: opWait, key: b.keyFrom(obj), waitEnd: make(chan bool, 1)}
+	timeout := time.NewTimer(b.MaxBatchPeriod)
+	select {
+	case b.ops <- waitBatchOp:
+		<-waitBatchOp.waitEnd
+	// if the ops channel is full (should be very rare), allow wait to block until the MaxBatchPeriod
+	case <-timeout.C:
+	}
+}
+
+// monitor is a synchronous loop that controls the window start, update, and end
+// monitor should be executed in one go routine and will handle all object batch windows
+func (b *Batcher) monitor(ctx context.Context) {
+	defer func() { b.isMonitorRunning = false }()
+	ticker := time.NewTicker(b.IdlePeriod / 2)
+	for {
+		select {
+		// Wake and check for any timed out batch windows
+		case <-ticker.C:
+			for key, batch := range b.windows {
+				b.checkForWindowEndAndNotify(key, batch)
+			}
+		// Process window operations
+		case op := <-b.ops:
+			switch op.kind {
+			// Start a new window or update progress on a window
+			case opAdd:
+				b.startOrUpdateWindow(op.key)
+			// Register a waiter and start a window if no window has been started
+			case opWait:
+				window, ok := b.windows[op.key]
+				if !ok {
+					window = b.startOrUpdateWindow(op.key)
+				}
+				window.closed = append(window.closed, op.waitEnd)
+			}
+		// Stop monitor routine on shutdown
+		case <-ctx.Done():
+			for key, window := range b.windows {
+				b.endWindow(key, window)
+			}
+			return
+		}
+	}
+}
+
+// checkForWindowEndAndNotify checks if a window has timed out due to inactivity (IdlePeriod) or has reached the MaxBatchPeriod.
+// If the batch window has ended, then the batch closed channel will be notified and the window will be removed
+func (b *Batcher) checkForWindowEndAndNotify(key string, window *window) {
+	if time.Since(window.lastUpdated) < b.IdlePeriod && time.Since(window.started) < b.MaxBatchPeriod {
+		return
+	}
+	b.endWindow(key, window)
+}
+
+// endWindow signals the end of a window to all wait consumers and deletes the window
+func (b *Batcher) endWindow(key string, window *window) {
+	for _, end := range window.closed {
+		select {
+		case end <- true:
+			close(end)
+		default:
+		}
+	}
+	delete(b.windows, key)
+}
+
+// startOrUpdateWindow starts a new window for the object key if one does not already exist
+// if a window already exists for the object key, then the lastUpdate time is set
+func (b *Batcher) startOrUpdateWindow(key string) *window {
+	batchWindow, ok := b.windows[key]
+	if !ok {
+		batchWindow = &window{lastUpdated: time.Now(), started: time.Now()}
+		b.windows[key] = batchWindow
+		return batchWindow
+	}
+	batchWindow.lastUpdated = time.Now()
+	if batchWindow.started.IsZero() {
+		batchWindow.started = time.Now()
+	}
+	return batchWindow
+}
+
+// keyFrom takes an object and outputs a unique key
+func (b *Batcher) keyFrom(obj metav1.Object) string {
+	return fmt.Sprintf("%s/%s", obj.GetName(), obj.GetNamespace())
```

Comment:
Thoughts on using selflink or id instead? This will avoid some niche collision scenarios.

Reply:
selflink is deprecated and is removed in 1.21 it looks like, but I think you're right that UID is probably a better choice here 👍